src/core/tsi/ssl/key_logging/ssl_key_logging.h

```diff
@@ -0,0 +1,149 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+#ifndef GRPC_CORE_TSI_SSL_KEY_LOGGING_SSL_KEY_LOGGING_H
+#define GRPC_CORE_TSI_SSL_KEY_LOGGING_SSL_KEY_LOGGING_H
+
+#include <grpc/support/port_platform.h>
+
+#include <iostream>
+
+#include <grpc/grpc_security.h>
+#include <grpc/slice.h>
+#include <grpc/support/sync.h>
+
+extern "C" {
+#include <openssl/ssl.h>
+}
+
+#include "src/core/lib/gprpp/memory.h"
+#include "src/core/lib/gprpp/ref_counted.h"
+#include "src/core/lib/gprpp/sync.h"
+
+/// Instance to facilitate loging of SSL/TLS session keys to aid debugging.
+///
+/// Keys logged by an instance of this class help decrypting packet captures
+/// with tools like wireshark.
+///
+/// This class is thread safe and serializes access to keylog files.
+
+namespace tsi {
+
+/// Configuration for key logging.
+struct tsi_tls_key_logger_config {
+  /// The full path at which the TLS keys would be exported.
+  std::string tls_key_log_file_path;
+  /// The format in which the TLS keys would be exported.
+  grpc_tls_key_log_format key_logging_format;
+
+  /// Future extensions can include filters such as IP addresses etc.
+
+  /// Constructor.
+  explicit tsi_tls_key_logger_config()
+      : tls_key_log_file_path(""),
+        key_logging_format(grpc_tls_key_log_format::TLS_KEY_LOG_FORMAT_NONE){};
+
+  /// Copy ctor.
+  tsi_tls_key_logger_config(const tsi_tls_key_logger_config& copy) {
+    tls_key_log_file_path = copy.tls_key_log_file_path;
+    key_logging_format = copy.key_logging_format;
+  }
+};
+
+/// A class which facilitates logging Tls keys into a file. The instance is
+/// bound to a file.
+class TlsKeyLogger : public grpc_core::RefCounted<TlsKeyLogger> {
+ public:
+  /// Instantiates a TlsKeyLogger instance bound to a specific path.
+  explicit TlsKeyLogger(const std::string& tls_key_log_file_path);
+  ~TlsKeyLogger() override;
+
+  /// Not copyable nor assignable.
+  TlsKeyLogger(const TlsKeyLogger&) = delete;
+  TlsKeyLogger& operator=(const TlsKeyLogger&) = delete;
+
+  /// Writes session keys into the file in the key logging format specified.
+  /// The passed string may be modified and logged based on the specified
+  /// format.
+  /// This is called upon completion of a handshake. The associated ssl_context
+  /// is also provided here to support future extensions such as logging
+  /// keys only when connections are made by certain IPs etc.
+  void LogSessionKeys(SSL_CTX* ssl_context,
+                      const tsi_tls_key_logger_config& tls_key_log_config,
+                      const std::string& session_keys_info);
+
+ private:
+  FILE* fd_;
+  grpc_core::Mutex lock_;
+  std::string tls_key_log_file_path_;
+};
+
+/// A Wrapper class which enables key logging to the a file based on specified
+/// configuration. A unique KeyLogger Container is bound to each Tls
+/// security connector.
+class TlsKeyLoggerContainer
```

Comment:
I am a bit confused on why we need a `TlsKeyLoggerContainer` in addition to `TlsKeyLogger`...can we make them into one class?

Reply:
I have renamed the classes here for more clarity. There are two classes: TlsKeyLogFileWriter and TlsKeyLogger. TlsKeyLogger is bound to a tls key log configuration i.e on object of TlsKeyLogger exists for each specified configuration. On the other hand one instance of TlsKeyLogFileWriter exists for each unique log file. So a two TlsKeyLoggers may share the same TlsKeyLogFileWriter instance if their key log configuration specifies the same log file. Having two separate classes like this allows the key log configuration (with addition of new entries) to grow over time while the TlsKeyLogFileWriter class can remain unchanged.